### PR TITLE
[SYCL-MLIR] Lower MemRef to pointers in -convert-sycl-to-llvm

### PR DESCRIPTION
**Pull request by victor-eds**
Populate converter with patterns exposed by polygeist.

### CODE REVIEW

File: mlir-sycl/lib/Conversion/SYCLToLLVM/CMakeLists.txt [hunk 1]
```diff
@@ -13,10 +13,6 @@ add_mlir_conversion_library(MLIRSYCLToLLVM
...
-  MLIRIR
-  MLIRLLVMCommonConversion
-  MLIRLLVMDialect
-  MLIRMemRefToLLVM
-  MLIRSYCLDialect  
-  MLIRTransforms
-  )
+  MLIRPolygeistTransforms
+  MLIRSYCLDialect
+)
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/CMakeLists.txt:18]: Cleaned and added `MLIRPolygeistTransforms`

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp [hunk 2]
```diff
@@ -55,6 +56,9 @@ void ConvertSYCLToLLVMPass::runOnOperation() {
   sycl::populateSYCLToLLVMConversionPatterns(converter, patterns);
   populateFuncToLLVMConversionPatterns(converter, patterns);
 
+  if (useBarePtrCallConv)
+    polygeist::populateBareMemRefToLLVMConversionPatterns(converter, patterns);
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp:60]: Actual change

File: polygeist/lib/Conversion/PolygeistToLLVM/CMakeLists.txt [hunk 3]
```diff
@@ -11,5 +11,10 @@ add_mlir_conversion_library(MLIRPolygeistToLLVM
   Core
 
   LINK_LIBS PUBLIC
+  MLIRMathToLLVM
+  MLIROpenMPToLLVM
+  MLIRPolygeist
+  MLIRSYCLToLLVM
+  MLIRVectorToLLVM
```
Comment (victor-eds) [inline polygeist/lib/Conversion/PolygeistToLLVM/CMakeLists.txt:18]: Only needed

File: polygeist/lib/Dialect/Polygeist/Transforms/CMakeLists.txt [hunk 4]
```diff
@@ -30,17 +30,13 @@ add_mlir_dialect_library(MLIRPolygeistTransforms
...
   MLIRNVVMDialect
-  MLIROpenMPToLLVM
   MLIRPass
   MLIRPolygeist
   MLIRSideEffectInterfaces
   MLIRSCFToControlFlow
-  MLIRSYCLDialect
-  MLIRSYCLToLLVM
   MLIRTransformUtils
-  MLIRVectorToLLVM
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/CMakeLists.txt:45]: Also cleaned

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir [hunk 5]
```diff
@@ -72,7 +72,7 @@ func.func @test_accessor.3(%arg0: !sycl_accessor_1_f32_rw_gb) {
 func.func @test_accessor.4(%arg0: !sycl_accessor_2_f32_rw_gb) {
   return
 }
-// CHECK: llvm.func @test_accessor.5(%arg0: !llvm.struct<"class.sycl::_V1::accessor{{.*}}", (struct<"class.sycl::_V1::local_accessor_base{{.*}}", ([[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]], struct<(ptr<i32, 3>
+// CHECK: llvm.func @test_accessor.5(%arg0: !llvm.struct<"class.sycl::_V1::accessor{{.*}}", (struct<"class.sycl::_V1::local_accessor_base{{.*}}", ([[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[RANGE_1]][[ARRAY_1]][[SUFFIX]], [[ID_1]][[ARRAY_1]][[SUFFIX]][[SUFFIX]], ptr<i32, 3>
```
Comment (victor-eds) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:75]: As we can see, types change here. We might want to make this option the default.

**Comment by victor-eds** [comment]
Depends on #8355

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir (hunk 5 above)
Comment (whitneywhtsang) [reply mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:75]: Agree to have this option on by default.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp [hunk 6]
```diff
@@ -52,9 +55,20 @@ void ConvertSYCLToLLVMPass::runOnOperation() {
...
+    populateReturnOpTypeConversionPattern(patterns, converter);
+    populateCallOpTypeConversionPattern(patterns, converter);
+    populateAnyFunctionOpInterfaceTypeConversionPattern(patterns, converter);
+  }
+
   sycl::populateSYCLToLLVMConversionPatterns(converter, patterns);
   populateFuncToLLVMConversionPatterns(converter, patterns);
 
+  if (useBarePtrCallConv)
+    polygeist::populateBareMemRefToLLVMConversionPatterns(converter, patterns);
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp:70]: These are the relevant changes. Changes to CMake files will have to be adapted after #8357 is merged.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir (hunk 5 above)
Comment (etiotto) [reply mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:75]: I think is essential that external functions use the bare pointer calling convention to be compatible with calls issued by other compilers. Agree it doesn't have to be done in this PR.

**Review by etiotto — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir (hunk 5 above)
Comment (victor-eds) [reply mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:75]: Fine, adding to my TODO list...

**Pull request merged**